client/txn.go

```diff
@@ -384,47 +378,40 @@ func (txn *Txn) exec(retryable func(txn *Txn) error) error {
 // always commit or clean-up explicitly even when that may not be
 // required (or even erroneous).
 func (txn *Txn) send(calls ...proto.Call) error {
+	if txn.Proto.Status != proto.PENDING {
+		return util.Errorf("attempting to use %s transaction", txn.Proto.Status)
+	}
+
 	lastIndex := len(calls) - 1
 	if lastIndex < 0 {
 		return nil
 	}
 
+	lastReq := calls[lastIndex].Args
 	// haveTxnWrite tracks intention to write. This is in contrast to
 	// txn.Proto.Writing, which is set by the coordinator when the first
 	// intent has been created, and which lives for the life of the
 	// transaction.
-	var haveTxnWrite bool
-	var haveEndTxn bool
+	haveTxnWrite := proto.IsTransactionWrite(lastReq)
 
-	for i, call := range calls {
+	for _, call := range calls[:lastIndex] {
```

Comment:
Whichever level is constructing the EndTransaction calls. If the EndTransaction call is alone in its batch, then it's easy: you can ask the Transaction object whether an EndTransaction is needed (i.e. inspect Transaction.Writing). If the EndTransaction is in the same batch as other calls, then it's trickier, but once we have end-to-end batching I'm not sure it's even necessary. Just include the EndTransaction in the batch whether it's necessary or not; the marginal cost on the server is small once it's no longer an independent raft command.